sql/plan/alter_table.go

```diff
@@ -418,6 +447,179 @@ func (i addColumnIter) Close(context *sql.Context) error {
 	return nil
 }
 
+// rewriteTable rewrites the table given if required or requested, and returns the whether it was rewritten
+func (i *addColumnIter) rewriteTable(ctx *sql.Context, rwt sql.RewritableTable) (bool, error) {
+	//return false, nil
+
+	newSch, projections, err := addColumnToSchema(i.a.targetSch, i.a.column, i.a.order)
+	if err != nil {
+		return false, err
+	}
+
+	oldPkSchema, newPkSchema := sql.SchemaToPrimaryKeySchema(rwt, rwt.Schema()), sql.SchemaToPrimaryKeySchema(rwt, newSch)
+
+	rewriteRequired := false
+	if i.a.column.Default != nil || !i.a.column.Nullable {
+		rewriteRequired = true
+	}
+
+	rewriteRequested := rwt.ShouldRewriteTable(ctx, oldPkSchema, newPkSchema, i.a.column)
+	if !rewriteRequired && !rewriteRequested {
+		return false, nil
+	}
+
+	inserter, err := rwt.RewriteInserter(ctx, newPkSchema)
+	if err != nil {
+		return false, err
+	}
+
+	partitions, err := rwt.Partitions(ctx)
+	if err != nil {
+		return false, err
+	}
+
+	rowIter := sql.NewTableRowIter(ctx, rwt, partitions)
+
+	for {
+		r, err := rowIter.Next(ctx)
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			return false, err
+		}
+
+		newRow, err := ProjectRow(ctx, projections, r)
+		if err != nil {
+			return false, err
+		}
+
+		err = inserter.Insert(ctx, newRow)
+		if err != nil {
+			return false, err
+		}
+	}
+
+	// TODO: move this into iter.close, probably
+	err = inserter.Close(ctx)
+	if err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+// addColumnToSchema returns a new schema and a set of projection expressions that when applied to rows from the old
+// schema will result in rows in the new schema.
+func addColumnToSchema(schema sql.Schema, column *sql.Column, order *sql.ColumnOrder) (sql.Schema, []sql.Expression, error) {
```

Comment:
this would be great in the analyzer, FixFieldIndexesForExpressions screws that up?

Reply:
It definitely belongs in the analyzer. My goal is to put everything there when I'm done with this round of changes.